enhancements/microshift/microshift-router-configuration.md

```diff
@@ -0,0 +1,530 @@
+changing nature of IP addresses, MicroShift will start a watcher for the listed
+NICs to update the service `status` as needed to minimize configuration drift.
+
+#### Allowing specific IP addresses
+Allowing specific IP addresses is achieved through the use of `NetworkPolicy`
+resources, as seen in previous examples. These policies include the allowed IP
+addresses and ports in the spec, meaning they reject everything else that is
+not included explicitly.
+
+Due to MicroShift's networking setup, there is one more change required for
+the `NetworkPolicy` to filter out based on the source IP.
+
+As it can be seen in the iptables rules, traffic towards the router ports is
+subject to DNAT to the router service IP and port. All services use
+`externalTrafficPolicy: Cluster`, which will SNAT traffic's source IP to match
+that of the node. This gets further NAT, as described here. By the time the
+`NetworkPolicy` is evaluated the source IP has been replaced with
+`100.64.0.0/16`, which will not produce matches against the configured IP in
+the policy, and traffic will get rejected.
+
+To overcome this, all LoadBalancer services in MicroShift that are subject to
+get NetworkPolicy resources to control external traffic must set
+`externalTrafficPolicy: Local`. This setting will not SNAT packets and the
+source IP is preserved until NetworkPolicy checks run.
+
+This setting, however, comes at the expense of an even load when having more
+than one node, as traffic will not be sent to another node on the same port.
+This is not an issue in single node, which is MicroShift's use case and design
+principle.
+
+This setting is also the default in OCP when deploying in cloud, having a load
+balancer before the cluster. Manifests for both OCP and MicroShift are
+compatible in this regard.
+
+## Open Questions
+N/A
+
+## Test Plan
+All of the changes listed here will be included in the current e2e scenario
+testing harness in MicroShift.
+
+## Graduation Criteria
+Targeting GA for MicroShift 4.16 release.
+
+### Dev Preview -> Tech Preview
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+
+### Tech Preview -> GA
+- More testing (upgrade, downgrade)
+- Sufficient time for feedback
+- Available by default
+- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+### Removing a deprecated feature
+N/A
+
+## Upgrade / Downgrade Strategy
+In the previous implementation, using host ports, the router would listen on
+any interface (except loopback). This included some of the internal IP
+addresses from ovnk, such as `169.254.169.2`, which is the link local IP
+address that ovnk uses in `br-ex` for internal handling of flows.
+
+The approach described in this enhancement does not expose the router in such
+IP, as it introduces conflicts between iptables rules and openflow.
```

Comment:
Maybe helpful to call out the firewalld issue here.  If I understand correctly, the cluster admin will need to do a manual migration by translating any custom firewalld rules into networkpolicy—right?

Reply:
Context for this section is a bit different. In current implementation the router is exposed in all of the interfaces, but not all of them work because of networking details. Link local address conflicts with some ovnk rules that take it into an endless loop of NATing, for example.
As for firewalld, in current implementation we do not have firewalld rules applying to the router, as hostPort rules take precedence over firewall. Since this approach is also outside of k8s control, network policies dont apply either. With this approach the firewall part is the same, but you can use network policies. This is a new capability that MicroShift didnt have before.

Reply:
Ah, when reading the "Firewalling ports" section, I thought the implication was that firewalld was the currently supported means of restricting access.  However, the "Why LoadBalancer service" section does that "the current setup" (i.e. without this enhancement) uses host ports, which prevents the use of firewalld to restrict access.  Thanks for clarifying!